flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java

```diff
@@ -209,20 +217,25 @@ public void registerTempCatalogScalarFunction(ObjectIdentifier oi, ScalarFunctio
 	private Set<String> getUserDefinedFunctionNames() {
 		Set<String> result = new HashSet<>();
 
-		// Get functions in catalog
+		// add temp system functions
+		result.addAll(tempSystemFunctions.keySet());
+
+		// add temp catalog functions
+		result.addAll(tempCatalogFunctions.keySet().stream()
+			.filter(oi -> oi.getCatalogName().equals(catalogManager.getCurrentCatalog())
+				&& oi.getDatabaseName().equals(catalogManager.getCurrentDatabase()))
+			.map(oi -> oi.getObjectName())
+			.collect(Collectors.toSet())
+		);
+
+		// add catalog functions
 		Catalog catalog = catalogManager.getCatalog(catalogManager.getCurrentCatalog()).get();
 		try {
 			result.addAll(catalog.listFunctions(catalogManager.getCurrentDatabase()));
```

Comment:
Nit: it might be cleaner to have a local var for current cat/db so as to avoid catalogManager.getXXX().